### PR TITLE
[k8s] Add flag to disable ssh setup

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Requested by user. Adds a flag to disable SSH setup, which can take ~10s per pod.

Depending on degree of parallelism available in an environment and total number of pods to provisioning, this could add delays of anywhere between 10 seconds and n*10 seconds.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual tests with `sky launch` with SkyPilot image and custom image
```
resources:
  cloud: kubernetes
  cpus: 1

num_nodes: 100

experimental:
  config_overrides:
    kubernetes:
      disable_ssh: true
```

### CODE REVIEW

File: sky/provision/kubernetes/instance.py [hunk 1]
```diff
@@ -668,7 +698,14 @@ def _create_pods(region: str, cluster_name_on_cloud: str,
         # on most base images. E.g., do not use Python, since that may not
         # be installed by default.
         _check_user_privilege(namespace, context, uninitialized_pods_list)
-        _setup_ssh_in_pods(namespace, context, uninitialized_pods_list)
+        if not provider_config.get('disable_ssh', False):
+            _setup_ssh_in_pods(namespace, context, uninitialized_pods_list)
```
Comment (Michaelvll) [inline sky/provision/kubernetes/instance.py:702]: Just curious, if the parallelism is taking most of the time. Can we just move the ssh setup script in to our setup commands for kubernetes:
https://github.com/skypilot-org/skypilot/blob/0f0008d41ecf4919b169541fb86a4ec60d1485ab/sky/templates/kubernetes-ray.yml.j2#L413-L431

The parallelism will only happen once then.

File: sky/provision/kubernetes/instance.py [hunk 2]
```diff
@@ -823,18 +860,19 @@ def get_cluster_info(
...
+        assert head_pod_name is not None
+        runner = command_runner.KubernetesCommandRunner(
+            ((namespace, context), head_pod_name))
+        rc, stdout, stderr = runner.run(get_k8s_ssh_user_cmd,
+                                        require_outputs=True,
+                                        separate_stderr=True,
+                                        stream_logs=False)
+        _raise_command_running_error('get ssh user', get_k8s_ssh_user_cmd,
+                                     head_pod_name, rc, stdout + stderr)
+        ssh_user = stdout.strip()
```
Comment (Michaelvll) [inline sky/provision/kubernetes/instance.py:875]: If we decide to move the SSH setup into the setup commands, we can still keep this.

File: sky/provision/kubernetes/instance.py (hunk 1 above)
Comment (romilbhardwaj) [reply sky/provision/kubernetes/instance.py:702]: Ah yeah that's a good point - we can refactor SSH setup to happen in `setup_commands`, and that way we can also waiting for SSH ready. That'll help reduce some latency.

However, most of the latency is not from the overheads of running more parallel threads, but rather the actual steps themselves (apt update, apt install openssh-server and then the service restart).

That said, I agree we should do that refactor - do you think we can do it in a separate PR? I can file an issue for it.

**Comment by romilbhardwaj** [comment]
Added conditional checks for `apt update` and `apt install` to further reduce provisioning time.

## Breakdown of SSH provisioning time

Master:
Total 13s to setup SSH
* 8s to apt update
* 3s to install
* 1s for rest of the commands

This branch:
Total 1s to setup SSH
* 1s to setup SSH if openssh is already present.

## Comparing with master and different values of `disable_ssh`:

Tested with single node`sky launch -y -c test --cloud kubernetes --image-id  nvcr.io/nvidia/nemo:24.05.01`

Master: ~2min 2s
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  5.73s user 3.97s system 7% cpu 2:02.55 total
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  6.31s user 4.12s system 7% cpu 2:04.55 total
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  5.68s user 3.94s system 7% cpu 2:03.55 total

This branch, without disable SSH: ~1 min 56s
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  4.84s user 3.78s system 7% cpu 1:58.31 total
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  5.41s user 3.71s system 7% cpu 1:56.03 total
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  5.55s user 3.74s system 8% cpu 1:55.17 total

This branch, with disable SSH: ~ 1 min 52s
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  4.71s user 3.72s system 7% cpu 1:51.74 total
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  5.76s user 3.82s system 8% cpu 1:54.76 total
sky launch -y -c test --cloud kubernetes --image-id  task4.yaml  5.13s user 3.72s system 7% cpu 1:52.71 total

Given that single node provisioning time is reduces by ~4s when SSH is disabled, and ~10s compared to master (with larger improvements for multi-node), it might be worthwhile to keep the disable_ssh flag.

**Comment by romilbhardwaj** [comment]
Identified that `apt update` is the slowest operation. Refactored to run apt-update in container init args, so it is run in parallel from the start.  

Testing on nemo image: `sky launch -y -c test --num-nodes 100 --cloud kubernetes --image-id  nvcr.io/nvidia/nemo:24.05.01`

Master branch: 19:56.21 total

this branch, SSH enabled: 15:26.56 total

this branch, SSH disabled: 14:47.28 total

Given it's a 40 second overhead to wait for SSH on 100 nodes , it might not be worthwhile disabling SSH (and having the flag to disable SSH). I'll close this PR for now and open a new PR with optimizations.

**Pull request closed without merging**